modules/L15_emulated_checks_nmap.sh

```diff
@@ -0,0 +1,206 @@
+  export NMAP_PORTS_SERVICES=()
+
+  if [[ "$SYS_ONLINE" -eq 1 ]] && [[ "$TCP" == "ok" ]]; then
+    module_log_init "${FUNCNAME[0]}"
+    module_title "Nmap scans of emulated device."
+
+    pre_module_reporter "${FUNCNAME[0]}"
+
+    if [[ -n "$IP_ADDRESS_" ]]; then
+
+      check_live_nmap_basic "$IP_ADDRESS_"
+
+      MODULE_END=1
+    else
+      print_output "[!] No IP address found"
+    fi
+    write_log ""
+    write_log "[*] Statistics:${#NMAP_SERVICES[@]}"
+    module_end_log "${FUNCNAME[0]}" "$MODULE_END"
+  fi
+
+}
+
+check_live_nmap_basic() {
+  local IP_ADDRESS_="${1:-}"
+  local NMAP_RESULT_FILES=()
+  local NMAP_RESULTF=""
+  local NMAP_SERVICES=()
+  local S09_L15_CHECK=()
+  local S116_L15_CHECK=()
+  local SERVICE=""
+  local SERVICE_NAME=""
+  local TYPE=""
+  export NMAP_PORTS_SERVICES=()
+
+  sub_module_title "Nmap portscans for emulated system with IP $ORANGE$IP_ADDRESS_$NC"
+
+  cp "$LOG_DIR"/l10_system_emulation/nmap_emba_"$IP_ADDRESS_"*.gnmap "$LOG_PATH_MODULE" || true
+  cp "$LOG_DIR"/l10_system_emulation/nmap_emba_"$IP_ADDRESS_"*.nmap "$LOG_PATH_MODULE" || true
+
+  # find all Nmap results
+  mapfile -t NMAP_RESULT_FILES < <(find "$LOG_PATH_MODULE" -name "*.nmap")
+  write_csv_log "---" "service identifier" "version_detected" "csv_rule" "license" "static/emulation/nmap/nikto"
+
+  if [[ -v NMAP_RESULT_FILES[@] ]]; then
+    for NMAP_RESULTF in "${NMAP_RESULT_FILES[@]}"; do
+      print_output "[*] Found Nmap results $ORANGE$(basename "$NMAP_RESULTF")$NC:"
+      tee -a "$LOG_FILE" < "$NMAP_RESULTF"
+      print_ln
+    done
+  else
+    # if no Nmap results are found we initiate a scan
+    nmap -Pn -n -sSV -A --host-timeout 30m "$IP_ADDRESS_" -oA "$LOG_PATH_MODULE"/nmap-basic-"$IP_ADDRESS_" | tee -a "$LOG_FILE"
+  fi
+  print_ln
+
+  # extract only the service details from gnmap output file:
+  mapfile -t NMAP_SERVICES < <(grep "open" "$LOG_PATH_MODULE"/*.gnmap | cut -d: -f2- | sed s/'\t'/'\n\t'/g | sed s/'\/, '/'\n\t\t'/g | sed s/'Ports: '/'Ports:\n\t\t'/g | grep -v "/closed/\|filtered/" | grep -v "Host: \|Ports:\|Ignored State:\|OS: \|Seq Index: \|Status: \|IP ID Seq: \|^# " | sed 's/^[[:blank:]].*\/\///' | sed 's/\/$//g'| sort -u || true)
+  mapfile -t NMAP_PORTS_SERVICES < <(grep "open" "$LOG_PATH_MODULE"/*.nmap | cut -d: -f2- | awk '{print $1,$3}' | sort -u || true)
+  # extract cpe information like the following:
+  #	Service Info: OS: Linux; Device: WAP; CPE: cpe:/h:dlink:dir-300:2.14, cpe:/o:linux:linux_kernel, cpe:/h:d-link:dir-300
+  mapfile -t NMAP_CPE_DETECTION < <(grep -ah "Service Info: " "$LOG_PATH_MODULE"/*.nmap | grep -a "CPE: .*" | sort -u)
+
+  TYPE="Nmap scan (Scan info)"
+
+  if [[ "${#NMAP_CPE_DETECTION[@]}" -gt 0 ]]; then
+    for NMAP_CPES in "${NMAP_CPE_DETECTION[@]}"; do
+      NMAP_CPES=$(echo "$NMAP_CPES" | grep -o "cpe:.*")
+      # rewrite the string into an array:
+      # NMAP_CPES_ARR=( $(echo "$NMAP_CPES" | tr "," "\n") )
+      IFS=', ' read -r -a NMAP_CPES_ARR <<< "$NMAP_CPES"
```

Comment:
readarray won't work here?

Reply:
I don't know ...

Reply:
Okey, thanks for clarifying it. As far as I know, there is a custom delimiter parameter for readarray: `-d '*'`. But in this case you can leave it, because it won't change anything in functionality. Just wondered and had to ask out of curiosity.

Reply:
we can do something like this:

```
readarray -d ', ' -t NMAP_CPES_ARR < <(printf "%s" "$NMAP_CPES")
for NMAP_CPE in "${NMAP_CPES_ARR[@]}"; do
   NMAP_CPE=${NMAP_CPE/ /}
```

Does this makes more sense?

Reply:
changed

Reply:
I have no idea - we should check in the future which approach is the better one. I have no idea how those two differ in the background. So for now we should accept it and move on :D